M2/Macaulay2/packages/ReactionNetworks.m2

```diff
@@ -311,7 +311,7 @@ addComplex(String, ReactionNetwork) := (c,Rn) -> (
 addReaction = method()
 addReaction(String, ReactionNetwork) := (r,Rn) -> (
     r = repairReaction(removeWhitespace r, Rn.NullSymbol);
-    complexes := apply(separateRegexp("(-->)|(<--)|(<-->)|,", r), removeWhitespace);
+    complexes := apply(separateRegexp("<-->|(?<!<)-->|<--(?!>)|,", r), removeWhitespace);
```

Comment:
There's no point to these lookarounds, because if there *is* a "<" there, we do want to match, as specified in the first clause.  Just use `"(<-->)|(-->)|(<--)|,"`.

Reply:
Also, this example illustrates a breaking change in behavior.  This PR:
```
+ ./M2 --no-readline --print-width 92
Macaulay2, version 1.16.0.1
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases,
               MinimalPrimes, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i6 : regex("a|aa","aa")

o6 = {(0, 1)}

o6 : List
```


Previous release:
```
+ /Applications/Macaulay2-1.16/bin/M2 --no-readline --print-width 93
Macaulay2, version 1.16
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases,
               MinimalPrimes, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : regex("a|aa","aa")

o1 = {(0, 2)}

o1 : List
```

Reply:
Your regular expression allows for both options, doesn't it? It's a change, but what does it break?

Reply:
> There's no point to these lookarounds, because if there _is_ a "<" there, we do want to match, as specified in the first clause. Just use `"(<-->)|(-->)|(<--)|,"`.

Sure.

Reply:
> Your regular expression allows for both options, doesn't it? It's a change, but what does it break?

It gives a different answer, so it could break anything.

Reply:
> > Your regular expression allows for both options, doesn't it? It's a change, but what does it break?
> 
> It gives a different answer, so it could break anything.

And it broke the match in the ReactionNetworks package, which is why you had to fix it.

Reply:
> And it broke the match in the ReactionNetworks package, which is why you had to fix it.

It broke the match because the authors had an incorrect assumption about how regex works. Nowhere in either our documentation or the documentation for GNU regex it is guaranteed that the algorithm starts with the rightmost alternation in `"(-->)|(<--)|(<-->)|,"` and go left, but since our regex behaved that way, the authors assumed that was correct. The lookarounds I added would ensure that there's no assumption about the order being made, either left-to-right or right-to-left. But if you don't like that, we can just move `<-->` first.

Switching from one regex engine to another, change is inevitable. Just like change is inevitable if we switch from one factorization library to another and they sort the answers differently. If Boost::regex is providing the wrong answer, I will look into it. But if someone baked incorrect assumptions in their regular expression and now it doesn't work, it is not a breaking change, it is a bug that I'm fixing.

Reply:
> It gives a different answer, so it could break anything.

Your regular expression was incorrect for what you wanted. If you want either one or two `a` characters, you can use this:
```m2
regex("a{1,2}", "aa") -- {(0, 2)}
```
Further, if you want to specify whether you like a shorter match or a longer match, you can use:
```m2
regex("a{1,2}?", "aa") -- {(0, 1)}
regex("a{1,2}+", "aa") -- {(0, 2)}
```

Reply:
> ... the algorithm starts with the rightmost alternation in `"(-->)|(<--)|(<-->)|,"` and go left

A simple experiment shows that's not what it's doing:

```
+ /Applications/Macaulay2-1.16/bin/M2 --no-readline --print-width 92
Macaulay2, version 1.16
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases,
               MinimalPrimes, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : regex("a|aa","aa")

o1 = {(0, 2)}

o1 : List

i2 : regex("aa|a","aa")

o2 = {(0, 2)}

o2 : List
```

Reply:
> > ... the algorithm starts with the rightmost alternation in `"(-->)|(<--)|(<-->)|,"` and go left
> 
> A simple experiment shows that's not what it's doing:
> 

At 

https://www.regular-expressions.info/posix.html

concerning `a|b` we read

"The POSIX standard, however, mandates that the longest match be returned."

In scclib.c we specify the use of that standard with the RE_SYNTAX_POSIX_EXTENDED option.

Reply:
You're right, the assumption isn't the order, but that regex returns the "leftmost, longest" match, which is no longer the case in the ECMAScript flavor of regex. I discussed this choice in https://github.com/Macaulay2/M2/issues/1092#issuecomment-661651568. Switching the flavor definitely changes some outputs, but like I said, the output is still a correct match for the regular expression. Moreover, the new flavor gives you better syntax to clarify what matches you prefer.

Read this discussion about the benefits of using ECMAScript instead of POSIX: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2003/n1429.htm#syntax_discussion

Reply:
In that case, let's offer both, using an optional argument to select the flavor, with posix as the default for upward compatibility.